dds/DCPS/RTPS/Spdp.cpp

```diff
@@ -396,91 +396,102 @@ namespace {
 
 #ifndef DDS_HAS_MINIMUM_BIT
 void
-Spdp::update_location(const DCPS::RepoId& guid,
-                      OpenDDS::DCPS::ParticipantLocation mask,
-                      const ACE_INET_Addr& from)
+Spdp::enqueue_location_update_i(DiscoveredParticipantIter iter,
+                                OpenDDS::DCPS::ParticipantLocation mask,
+                                const ACE_INET_Addr& from)
 {
-   ACE_GUARD(ACE_Thread_Mutex, g, lock_);
-   update_location_i(guid, mask, from);
+  // We have the global lock.
+  iter->second.location_updates_.push_back(DiscoveredParticipant::LocationUpdate(mask, from));
 }
 
 void
-Spdp::update_location_i(const DCPS::RepoId& guid,
-                        OpenDDS::DCPS::ParticipantLocation mask,
-                        const ACE_INET_Addr& from)
+Spdp::process_location_updates_i(DiscoveredParticipantIter iter)
 {
   // We have the global lock.
-  DiscoveredParticipantIter iter = participants_.find(guid);
-  if (iter == participants_.end()) {
+
+  if (iter->second.bit_ih_ == DDS::HANDLE_NIL) {
+    // Not in the built-in topics.
     return;
   }
 
-  OpenDDS::DCPS::ParticipantLocationBuiltinTopicData& ld = iter->second.location_data_;
+  const RepoId guid = iter->first;
 
-  OPENDDS_STRING addr = "";
-  ACE_TCHAR buffer[256];
+  DiscoveredParticipant::LocationUpdateList location_updates = iter->second.location_updates_;
+  iter->second.location_updates_.clear();
```

Comment:
This looks like it could be a `swap` (with previous line), should be more efficient.